File: 0039-combination-sum/0039-combination-sum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> currentCombination = new ArrayList<>();
        
        Arrays.sort(candidates);
        dfs(candidates, target, 0, currentCombination, result);
        
        return result;
    }
    
    private void dfs(int[] candidates, int target, int start, List<Integer> currentCombination, List<List<Integer>> result) {
        if (target == 0) {
            result.add(new ArrayList<>(currentCombination));
            return;
        }
        
        for (int i = start; i < candidates.length && candidates[i] <= target; i++) {
            currentCombination.add(candidates[i]);
            dfs(candidates, target - candidates[i], i, currentCombination, result);
            currentCombination.remove(currentCombination.size() - 1);
        }
    }
}

File: 0039-combination-sum/0039-combination-sum.cs
public class Solution {
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        List<IList<int>> result = new List<IList<int>>();
        List<int> currentCombination = new List<int>();
        Array.Sort(candidates);
        dfs(candidates, target, 0, currentCombination, result);
        return result;
    }
    
    private void dfs(int[] candidates, int target, int start, List<int> currentCombination, List<IList<int>> result) {
        if(target == 0) {
            result.Add(new List<int>(currentCombination));
            return;
        }
        
        for(int i = start; i < candidates.Length && candidates[i] <= target; i++) {
            currentCombination.Add(candidates[i]);
            dfs(candidates, target - candidates[i], i, currentCombination, result);
            currentCombination.RemoveAt(currentCombination.Count - 1);
        }
    }
}
